### ISSUE TITLE
Continous Batching support with TGI and transformers-neuronx

### ISSUE THREAD

**Issue by mmcclean-aws** [huggingface/optimum-neuron#288]
Continuous Batching has been proven to improve throughput of LLMs and is requested by customers

**Comment by dacorvo** [comment]
Continuous batching is supported by TGI service, and available in the neuronx-tgi docker image.

**Comment by yinsong1986** [comment]
Hi @dacorvo, 

Can you point out to the document how to use  Continuous batching in the neuronx-tgi docker image? It seems like it is still using stack batching?

Thank you very much!

**Comment by yinsong1986** [comment]
@dacorvo I found the **TGI container cannot take concurrent requests larger than the model batch size**, do you know any work around for this? @dacorvo Thank you!

**Comment by dacorvo** [comment]
There might be a misunderstanding about what continuous batching is: this does not increase the number of concurrent requests, but instead allows to fill empty "slots" with new requests when some pending requests are finished.
Without continuous batching, a pending request would be queued until all request sin the batch have finished.
With continuous batching, it may leave the queue much faster if one of the requests completes before the others.

**Comment by yinsong1986** [comment]
@dacorvo Thanks for your patient explanation! Now I understand `the hard limit for concurrent requests is the batch size`. But some times we need to support **very high concurrent requests**? For example, the model batch size is 4, but it's required to support max concurrent requests equals to 512.

 Is there anything we can do to support this? Thanks a lot!

**Comment by dacorvo** [comment]
Without increasing the batch size, there are two options to increase the number of concurrent requests:

- increase the number of TGI instances per host and use SageMaker to do the load-balancing. We will create a detailed guide for this soon, but in the meantime you can take a look at: https://aws.amazon.com/de/blogs/aws/amazon-sagemaker-adds-new-inference-capabilities-to-help-reduce-foundation-model-deployment-costs-and-latency/,
- use multiple host and have a front-end to do the load balancing.

Now regarding the batch size, we might have good news soon: in the next release we might be able to support much higher batch sizes for Llama and Mistral models.

**Comment by yinsong1986** [comment]
Thanks @dacorvo That's really helpful!

BTW: Just another random idea :) Is it possible to add an internal buffer (or a queue to store pending requests) inside TGI neuron? So it can handle high concurrent requests without the need to increase the model batch size? Pls let me know your thoughts if possible. Thanks!

**Comment by dacorvo** [comment]
> BTW: Just another random idea :) Is it possible to add an internal buffer (or a queue to store pending requests) inside TGI neuron? So it can handle high concurrent requests without the need to increase the model batch size?

It is a design choice to immediately return a response when the server is full, i.e. when `max_concurrent_request` is exceeded. This is because there are plenty of different ways of addressing the queuing/load-balancing before actually reaching the TGI server.

cc @narsil

**Comment by yinsong1986** [comment]
Thanks for your quick response @dacorvo 

to the best of my understanding, the original TGI actually implemented such a queue inside the container https://github.com/huggingface/text-generation-inference/tree/v1.0.2/router, so the original TGI would support very high concurrency? Pls correct if I am wrong here :)

My opinionated view is it might be more natural to keep this feature of Neuron TGI consistent with the original TGI? So the users can have a consistent user experience:) Pls let me know your thoughts and thanks!

**Comment by Narsil** [comment]
`max_concurrent_requests` is the flag controlling the queue size, you can just increase it which will send more queries into the queue waiting.

Having early error return is what enables backpressure control and is essential to handle extremely high loads. 
Basically it's better to return an error if the request is going to take 10mn (users are likely gone by then). By returning early the server avoid the job of keeping the connection alive and treating it correctly, meaning when the load is gone, it will recover super fast.

We do kill connections when clients kill the connection, but users might just look at another tab or go somewhere else.

If you want to just queue a lot, just increase the max_concurrent_requests.

**Comment by yinsong1986** [comment]
Thanks for your reply! @Narsil

I have tried increasing the max_concurrent_requests to 128 when `batch_size=4`, but usually got a lot of errors like:

```
Request failed during generation: Server error: Cannot prefill 5 new request(s) with only 4 empty slots.Please align the number of concurrent requests with the static batch size: 4.
```
Was the above behaviour the expected? Thank you!

**Comment by dacorvo** [comment]
@yinsong1986 yes, this is expected, even if it is not ideal. As I said, even if you increase the `max_concurrent_request`, the hard limit is the `batch_size`. For now the information exchanged between the TGI server and the TGI front-end is not sufficient, leading to the latter being too optimistic when evaluating the actual capacity of the former. It will therefore try to push more requests that can be handled. 
This is why the current recommended configuration is `max_concurrent_requests == batch_size`.
We are working on addressing this issue.

**Comment by yinsong1986** [comment]
Thank you @dacorvo That makes a lot of sense and look forward to the work you are doing to address this. Cheers!

**Comment by yinsong1986** [comment]
Hi @dacorvo 

Sorry to to bother you again :) Not sure if you noticed, the original transformers-neuronx seems to have continuous batching supported here https://github.com/aws-neuron/transformers-neuronx/blob/main/src/transformers_neuronx/base.py#L250.

Is the above related? if it is related, do you have plan to integrate the above continuous batching into this TGI continuous batching implementation? Thank you!

**Comment by dacorvo** [comment]
@yinsong1986 we are back to the discussion a few days ago:

> There might be a misunderstanding about what continuous batching is: this does not increase the number of concurrent requests, but instead allows to fill empty "slots" with new requests when some requests are finished.

**Comment by dacorvo** [comment]
Implemented in 0.0.20
